### DISCUSSION TITLE
Input validation uses wrong error message when using callable class

### DISCUSSION THREAD

**Discussion by iron3oxide** [zauberzeug/nicegui#2333] [Q&A] [answered]
### Question

A bit of background:
I'm developing an app that provides both a UI and an API and I want the same validation functions I use for input validation in the UI to be used for Pydantic model validation of PATCH/PUT requests. I have thus written them in such a way that they return the given (or in some cases modified) value and throw an `AssertionError` if validation fails. In order to use them with `nicegui`, I have built two wrappers: one that bundles them with their corresponding error message (`InputValidator`), and another (`InputValidation`), that takes a list of these and when called with a value, calls the validation function with said value and returns the error message when an `AssertionError` occurs or `None` if the validation succeeded.

```python
import dataclasses
from typing import Any, Callable


@dataclasses.dataclass
class InputValidator:
    msg: str
    func: Callable[[Any], Any]


class InputValidation:
    def __init__(self, validators: list[InputValidator]) -> None:
        self.validators = validators
        self.value = None

    def __call__(self, value: Any) -> str | None:
        self.value = value
        for validator in self.validators:
            try:
                self.value = validator.func(self.value)
                if self.value is None:
                    return None  # positive early return
            except AssertionError:
                return validator.msg  # negative early return
        return None  # complete validation successful
```

Problem:
I can see that the validation works as the input turns red when I type in multiple kinds of invalid text. However, the error message always stays the same. I have debugged this and curiously enough, the right `validator.msg` is returned, but seemingly not used. Why might this be?

**Comment by iron3oxide** [comment]
A more complete and reproducible example:
```python
import dataclasses
from datetime import date, datetime
from typing import Any, Callable

from nicegui import ui


@dataclasses.dataclass
class InputValidator:
    msg: str
    func: Callable[[Any], Any]


class InputValidation:
    def __init__(self, validators: list[InputValidator]) -> None:
        self.validators = validators
        self.value = None

    def __call__(self, value: Any) -> str | None:
        self.value = value
        for validator in self.validators:
            try:
                self.value = validator.func(self.value)
                if self.value is None:
                    return None  # positive early return
            except AssertionError:
                return validator.msg  # negative early return
        return None  # complete validation successful


def required(input: str) -> str:
    assert input is not None and len(input) > 0
    return input


def valid_date_string(input: str) -> date:
    try:
        input_date = datetime.strptime(input, "%d.%m.%Y").date()

    except ValueError:
        raise AssertionError
    return input_date


def too_old(input: date) -> date:
    assert input > date.today()
    return input


RequiredStr = InputValidator(msg="Input required", func=required)
TooOld = InputValidator(msg="Too short.", func=too_old)
ValidDateStr = InputValidator(msg="Invalid date string.", func=valid_date_string)


def test_input() -> None:
    ui.input(
        validation=InputValidation(
            validators=[
                RequiredStr,
                ValidDateStr,
                TooOld,
            ]
        )
    )
```

**Comment by stepheweffie** [comment]
```py
import dataclasses
from datetime import date, datetime
from typing import Any, Callable
from nicegui import ui


@dataclasses.dataclass
class InputValidator:
    msg: str
    func: Callable[[Any], Any]


class InputValidation:
    def __init__(self, validators: list[InputValidator]) -> None:
        self.validators = validators
        self.value = None

    def __call__(self, value: Any) -> str | None:
        self.value = value
        for validator in self.validators:
            try:
                self.value = validator.func(self.value)
                if self.value is None:
                    return None  # positive early return
            except AssertionError:
                return validator.msg  # negative early return
        return None  # complete validation successful


def required(inp: str) -> str:
    assert inp is not None and len(inp) > 0
    return inp


def valid_date_string(inp: str) -> date:
    try:
        input_date = datetime.strptime(inp, "%d.%m.%Y").date()
    except ValueError:
        raise AssertionError
    return input_date


def too_old(inp: date) -> date:
    assert inp > date.today()
    return inp


RequiredStr = InputValidator(msg="Input required", func=required)
TooOld = InputValidator(msg="Too short.", func=too_old)
ValidDateStr = InputValidator(msg="Invalid date string.", func=valid_date_string)


def validate_input(inp: ui.input, validation: InputValidation) -> None:
    # Perform validation on the input value
    error_message = validation(inp.value)
    if error_message:
        ui.notify(error_message, color="warning")
    else:
        inp.value = validation.value
        ui.notify("Input valid", color="positive")


with ui.input():
    inp = ui.input(label='Date dd.mm.yyyy:')
    validation = InputValidation(validators=[RequiredStr, ValidDateStr, TooOld])
    ui.button('Validate', on_click=lambda: validate_input(inp, validation))


ui.run()
```

test this - it works

**Reply by iron3oxide** [reply]
Thank you for taking the time to test this! I have no doubt that it works, however it is not exactly what I intend to do, as I would like to use the inbuilt automatic validation functionality of `ui.input()`, which works like a charm when I e.g. use a dict and bool functions instead.

**Comment by iron3oxide** [comment, marked as answer]
I think I found the culprit! In line 28 of `validation_element.py`, the `error-message` prop is set via the `_props` dict instead of `.props()`, which would update the element. @falkoschindler @rodja is there a reason the latter isn't used there? If not, would it be okay if I converted this to an issue and made a PR? I'd also fix the type hint of `validation`, which doesn't include the `Callable` option atm

**Reply by falkoschindler** [reply]
Oh wow, that's awesome! I just [noticed](https://github.com/zauberzeug/nicegui/discussions/2333#discussioncomment-8082672) that there's a bug, but didn't find where exactly.

Using the `.props()` method is a bit tricky here, because you'd need to escape special characters before they will be parsed again. Writing the dictionary item directly doesn't need these two steps. But if we are already in the "error" state, the `self.props('error')` doesn't trigger an update. A possible solution would be to check if the error message changes and to explicitly call `.update()` in this case.

I also spottet the slightly wrong type annotation. And there are other elements derived from `ValidationElement` with the same problem. It would be great if you could fix them, once you're at it. 🙂

**Comment by falkoschindler** [comment]
Hi @iron3oxide!

I tried to fix your code, but it looks like there's (at least) a bug in the validation implementation in `ui.input`. Here's a minimal example with multiple possible validation errors:

```py
ui.input(validation=lambda value: '< 3' if len(value) < 3 else '< 5' if len(value) < 5 else None)
```

When you type "abcde", the error doesn't switch from "< 3" to "< 5". And when you delete the value character by character, the error doesn't switch from "< 5" to "< 3". 😕

**Reply by iron3oxide** [reply]
Thanks for the effort! I think I already know what causes this, see my reply above
